raids: scouter screenshotting

Alexsuperfly opened the pull request:
Displays the cc and world the scout is performed on
Adds the ability to screenshot only the scouting overlay for easy sharing with other raid members

![](https://user-images.githubusercontent.com/7191512/64051477-35f7e380-cb49-11e9-87a2-c10b53b6683f.png)

Refactors an ImageCapture utility class out of the Screenshot plugin in order for the Raids plugin to also utilize it

based slightly on #6092

Alexsuperfly commented:
Narrowed the scope of his PR by extracting the recommended items display to pr #9778
the items are subjective and were sort of tacked on to this without any true need to be coupled together

Nightfirecat commented:
A note aside from code review, I'm finding it hard to review this commit-by-commit because it's too split up. I would expect to see this PR's changes contained in far fewer commits. Something resembling the following is what I would expect to see:

1. `util: Add ImageCapture class`
2. `screenshot: Use ImageCapture to capture screenshots`
3. `raids: Enable screenshotting of scouting overlay`
4. `raids: Add clan chat and world to scouting overlay`

Generally speaking, outside of creating utility classes, each commit should cause some self-contained change in behavior to the client.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsConfig.java:
```diff
@@ -88,6 +90,17 @@ default boolean scoutOverlayInRaid()
 		position = 5,
+		keyName = "ccDisplayStyle",
+		name = "CC and World in scout overlay",
+		description = "Display style of current CC and World in scouting overlay"
```
There's not really a reason to abbreviate "clan chat" in the description. Let's write that out to be explicit and clear.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsOverlay.java:
```diff
@@ -98,6 +107,25 @@ public Dimension render(Graphics2D graphics)
+					.right(clanOwner)
+					.leftColor(Color.ORANGE)
+					.rightColor(color)
+					.build());
```
Each of these lines should have their indent reduced by one.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsPlugin.java:
```diff
@@ -78,7 +91,9 @@
 @Slf4j
 public class RaidsPlugin extends Plugin
 {
+	private static final DateFormat TIME_FORMAT = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss");
```
Instead of duplicating this across plugins which will share the screenshot utility, we could define this as a constant in the screenshot util.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsPlugin.java:
```diff
@@ -193,6 +216,17 @@ public void onVarbitChanged(VarbitChanged event)
+	{
+		if (event.getEntry().getMenuAction() == MenuAction.RUNELITE_OVERLAY &&
+				event.getEntry().getTarget().equals("Raids overlay") &&
+				event.getEntry().getOption().equals("Screenshot"))
```
Reduce indent by one for these lines.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsPlugin.java:
```diff
@@ -193,6 +216,17 @@ public void onVarbitChanged(VarbitChanged event)
+	@Subscribe
+	public void onOverlayMenuClicked(OverlayMenuClicked event)
+	{
+		if (event.getEntry().getMenuAction() == MenuAction.RUNELITE_OVERLAY &&
```
Code style throughout the project is to move operands to a new line.

```java
if (foo
	&& bar
	&& baz)
{
	// ...
}
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsPlugin.java:
```diff
@@ -625,4 +659,55 @@ private RaidRoom determineRoom(Tile base)
+			overlay.render(graphic);
+		}
+
+		graphic.setFont(FontManager.getRunescapeFont());
```
Can we get the user's selected font rather than defaulting to the RS font?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/util/ImageCapture.java:
```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) 2019, Alexsuperfly <https://github.com/Alexsuperfly>
```
While you can include your copyright on this since you adapted it, it should also include the copyright of the original author. (looks like lotto?)

Alexsuperfly commented:
i believe i have addressed everything, i forgot to use your commit message suggestions when making them again but its close enough.  

I did have to remove the only showing the CC and world in screenshot feature because it wouldnt send the dimensions correctly no matter what i tried if you used different fonts, but its not too big a loss

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/util/ImageCapture.java:
```diff
@@ -0,0 +1,218 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
There should be no newline between the copyright header and package definition.
```suggestion
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsOverlay.java:
```diff
@@ -98,6 +106,27 @@ public Dimension render(Graphics2D graphics)
+			color = Color.ORANGE;
+			FontMetrics metrics = graphics.getFontMetrics();
+			String worldString = "W" + client.getWorld();
+			String clanOwner = Text.removeTags(client.getWidget(WidgetInfo.CLAN_CHAT_OWNER).getText());
```
This should be done using `Client.getClanMemberManager()`.

Alexsuperfly commented:
Rebased and addressed your review. Thanks for pointing out `Client.getClanMemberManager()` I didn't realize that existed, its much better than what was being used before.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsOverlay.java:
```diff
@@ -101,6 +104,28 @@ public Dimension render(Graphics2D graphics)
+			ClanMemberManager clanMemberManager = client.getClanMemberManager();
+			FontMetrics metrics = graphics.getFontMetrics();
+			String worldString = "W" + client.getWorld();
+			String clanOwner = "Open CC tab...";
```
I don't think this is dependent upon the tab being open?

Alexsuperfly replied:
correct, its a relic of the method that was being used before clanMemeberManager existed, ill just make it "Join a CC" or something.

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsPlugin.java:
```diff
@@ -340,6 +359,12 @@ public void onOverlayMenuClicked(final OverlayMenuClicked event)
 		{
 			sendRaidLayoutMessage();
 		}
+		else if (event.getEntry().getMenuAction() == MenuAction.RUNELITE_OVERLAY
```
Two of these conditions can be pulled out and combined with the above one

Adam- commented on runelite-client/src/main/java/net/runelite/client/plugins/raids/RaidsPlugin.java:
```diff
@@ -752,4 +777,32 @@ private boolean submitRaid(ChatInput chatInput, String s)
 		return true;
 	}
+
+	private final HotkeyListener hotkeyListener = new HotkeyListener(() -> config.hotkey())
```
This and the config opt should probably be `screenshotHotkey`?
